readthedocs/projects/tasks.py

```diff
@@ -715,6 +723,123 @@ def save_build_config(self):
+            pdf=False,
+            epub=False,
+    ):
+        """
+        Save build artifacts to "storage" using Django's storage API
+
+        The storage could be local filesystem storage OR cloud blob storage
+        such as S3, Azure storage or Google Cloud Storage.
+
+        Remove build artifacts of types not included in this build (PDF, ePub, zip only).
+
+        This looks very similar to `move_files` and is intended to replace it!
+
+        :param html: whether to save HTML output
+        :param localmedia: whether to save localmedia (htmlzip) output
+        :param search: whether to save search artifacts
+        :param pdf: whether to save PDF output
+        :param epub: whether to save ePub output
+        """
+        if settings.RTD_BUILD_MEDIA_STORAGE:
+            log.info(
+                LOG_TEMPLATE.format(
+                    project=self.version.project.slug,
+                    version=self.version.slug,
+                    msg='Writing build artifacts to media storage',
+                ),
+            )
+
+            storage = get_storage_class(settings.RTD_BUILD_MEDIA_STORAGE)()
+
+            types_to_copy = []
+            types_to_delete = []
+
+            # HTML media
+            if html:
+                types_to_copy.append(('html', self.config.doctype))
+
+            # Search media (JSON)
+            if search:
+                types_to_copy.append(('json', 'sphinx_search'))
+
+            if localmedia:
+                types_to_copy.append(('htmlzip', 'sphinx_localmedia'))
+            else:
+                types_to_delete.append('htmlzip')
+
+            if pdf:
+                types_to_copy.append(('pdf', 'sphinx_pdf'))
+            else:
+                types_to_delete.append('pdf')
+
+            if epub:
+                types_to_copy.append(('epub', 'sphinx_epub'))
+            else:
+                types_to_delete.append('epub')
+
+            for media_type, build_type in types_to_copy:
+                from_path = self.version.project.artifact_path(
+                    version=self.version.slug,
+                    type_=build_type,
+                )
+                to_path = self.version.project.get_storage_path(
+                    type_=media_type,
+                    version_slug=self.version.slug,
+                    include_file=False,
+                )
+                log.info(
+                    LOG_TEMPLATE.format(
+                        project=self.version.project.slug,
+                        version=self.version.slug,
+                        msg=f'Writing {media_type} to media storage - {to_path}',
+                    ),
+                )
+                try:
+                    storage.copy_directory(from_path, to_path)
+                except Exception:
+                    # Ideally this should just be an IOError
+                    # but some storage backends unfortunately throw other errors
+                    log.warning(
```

Comment:
Now that we have we are duplicating the copy a warning could be enough. Although, I think we want to log an `.exception` here so we can see it under Sentry and take care of it. Otherwise, if sync fails for any reason we won't know that and we will see random issues on webpages.

Reply:
Anyway, `.exception` could generate a flood of notifications here.

Reply:
Maybe a `log.warning` with `exc_info=True`?

Reply:
I'm not sure what I want :)

My concern is that if this fail we won't be alarmed/notified at all. Although, alarming us for every single file that was not able to be copied is crazy.

Reply:
I think we just do a `log.exception` and Sentry can sort it out. It looks like Sentry is smart enough to find similar issues even when the text isn't exactly the same.